Update matrix-stats aggregation. (#9434)
opensearch-project/documentation-website: _aggregations/metric/matrix-stats.md

# Matrix stats aggregations

> The `matrix_stats` metric is a multi-value metric that generates covariance statistics for two or more fields in matrix form. 

Author: 1. Does `matrix-stats` always sample every available document? If not, how is sampling done?

2. What assumptions does `matrix-stats` make about independence of data samples? Is it sufficent to say that it assumes documents are independent of each other?

Reviewer: Again will probably keep it as `multi-value metric aggregation` and maybe link to the metric aggregation page

Author: @jainankitk Please see my two technical questions above. Can you provide any insight?

Reviewer 2: @jainankitk Could you please answer this quiestion?

Reviewer: > Does matrix-stats always sample every available document? If not, how is sampling done?

I don't think there is any sampling, all the documents are processed

> What assumptions does matrix-stats make about independence of data samples? Is it sufficent to say that it assumes documents are independent of each other?

`Matrix-stats` does not make any assumptions about the data. I am wondering why does it need to make any assumption?

Author: @jainankitk There are several assumptions that have to be met for a correlation to be valid: https://www.statology.org/pearson-correlation-assumptions/. However, that level of detail is probably beyond the scope of the documentation here. At some level, the user is assumed to know what they're doing if they use an aggregation.

> The `matrix_stats` metric is a multi-value metric that generates covariance statistics for two or more fields in matrix form. 

Reviewer 2: ```suggestion
The `matrix_stats` metric is a multi-value metric aggregation that generates covariance statistics for two or more fields in matrix form. 
```

> The `matrix_stats` metric does not support scripting.

Reviewer: aggregation does not support scripting?

Author: @jainankitk That's my understanding. Please verify.

Reviewer: Okay, my point was that the below reads slightly better?

```
The `matrix_stats` aggregation does not support scripting.
```

Author: ```suggestion
The `matrix_stats` aggregation does not support scripting.
```

{: .note}

## Parameters

The `matrix_stats` aggregation takes the following parameters: 

> | Parameter | Required/Optional | Data type      | Description |

Author: I tested the `missing` parameter and it gave me an error. Does `missing` not work with `matrix-stats`?

Reviewer 2: @jainankitk Could you please answer this quiestion?

Author: @jainankitk Maybe `missing` doesn't work because it takes an array of input fields? I tried the following and got an error:
```
GET opensearch_dashboards_sample_data_ecommerce/_search
{
  "size": 0,
  "aggs": {
    "matrix_stats_taxful_total_price": {
      "matrix_stats": {
        "fields": ["taxful_total_price", "products.base_price"],
        "missing": [0, 0]
      }
    }
  }
}
```

Reviewer: Okay you need to provide map of fieldname:value in the missing parameter, like below:

```
{
  "size": 0,     
  "aggs": {
    "matrix_stats_taxful_total_price": {
      "matrix_stats": {
        "fields": ["status", "size"],
        "missing": {"status":0}
}}}}
```

| :--       | :--               | :--            | :--         |
| `fields`  | Required          | String         | An array of fields for which the matrix stats are computed. |

> | `mode`    | Optional          | String         | Which value to use as a sample from a multi-valued or array field. Allowed values are `avg`, `min`, `max`, `sum`, and `median`.  |

Reviewer 2: @jainankitk Could you please answer this quiestion?

> | `mode`    | Optional          | String         | Which value to use as a sample from a multi-valued or array field. Allowed values are `avg`, `min`, `max`, `sum`, and `median`.  |

Author: @jainankitk Try the following:

```
DELETE /students

POST _bulk
{ "create": { "_index": "students", "_id": "1" } }
{ "name": "John Doe", "gpa": 3.89, "class_grades": [3.0, 3.9, 4.0], "grad_year": 2022}
{ "create": { "_index": "students", "_id": "2" } }
{ "name": "Jonathan Powers", "gpa": 3.85, "class_grades": [4.1, 3.0, 4.0], "grad_year": 2025 }
{ "create": { "_index": "students", "_id": "3" } }
{ "name": "Jane Doe", "gpa": 3.52, "class_grades": [3.2, 2.1, 3.8], "grad_year": 2024 }

GET students/_search
{
  "size": 0,
  "aggs": {
    "matrix_stats_taxful_total_price": {
      "matrix_stats": {
        "fields": ["gpa", "class_grades"],
        "mode": "avg"
      }
    }
  }
}
```

In all cases, the correlation between `gpa` and `class_grades` is the same, 0.9820867239098596. Surely it should be different if the value used for `class_grades` is computed differently for different `mode` values?

> | `mode`    | Optional          | String         | Which value to use as a sample from a multi-valued or array field. Allowed values are `avg`, `min`, `max`, `sum`, and `median`.  |

Reviewer: Thanks @dwelsch-esi for close observation. I noticed that the first response is cached resulting in this discrepancy. You can try using the `GET students/_search?request_cache=false` parameter to avoid this issue. Or alternatively, `POST /_cache/clear` before each search request

> | `mode`    | Optional          | String         | Which value to use as a sample from a multi-valued or array field. Allowed values are `avg`, `min`, `max`, `sum`, and `median`.  |

Reviewer: we have PR for fixing this discrepancy - https://github.com/opensearch-project/OpenSearch/pull/18254, in case anyone is curious

# Matrix stats aggregations

The `matrix_stats` aggregation is a multi-value metric aggregation that generates covariance statistics for two or more fields in matrix form. 

The `matrix_stats` aggregation does not support scripting.
{: .note}

## Parameters

The `matrix_stats` aggregation takes the following parameters. 

| Parameter | Required/Optional | Data type      | Description |
| :--       | :--               | :--            | :--         |
| `fields`  | Required          | String         | An array of fields for which the matrix stats are computed. |
| `missing`    | Optional          | Object         | The value to use in place of missing values. By default, missing values are ignored. See [Missing values](#missing-values). |
| `mode`    | Optional          | String         | The value to use as a sample from a multi-valued or array field. Allowed values are `avg`, `min`, `max`, `sum`, and `median`. Default is `avg`. |

## Example

> This example returns statistics for the `taxful_total_price` and `products.base_price` fields in the ecommerce sample data:

Reviewer 3: ```suggestion
The following example returns statistics for the `taxful_total_price` and `products.base_price` fields in the OpenSearch Dashboards e-commerce sample data:
```

    "hits": []
  },
  "aggregations": {
    "matrix_stats_taxful_total_price": {
      "doc_count": 4675,
      "fields": [
        {
          "name": "products.base_price",
          "count": 4675,
          "mean": 34.99423943014724,
          "variance": 360.5035285833702,
          "skewness": 5.530161335032689,
          "kurtosis": 131.1630632404217,
          "covariance": {
            "products.base_price": 360.5035285833702,
            "taxful_total_price": 846.6489362233169
          },
          "correlation": {
            "products.base_price": 1,
            "taxful_total_price": 0.8444765264325269
          }
        },
        {
          "name": "taxful_total_price",
          "count": 4675,
          "mean": 75.05542864304839,
          "variance": 2788.1879749835425,
          "skewness": 15.812149139923994,
          "kurtosis": 619.1235507385886,
          "covariance": {
            "products.base_price": 846.6489362233169,
            "taxful_total_price": 2788.1879749835425
          },
          "correlation": {
            "products.base_price": 0.8444765264325269,
            "taxful_total_price": 1
          }
        }
      ]
    }
  }
}
```

> The following table describes the response fields:

Reviewer: describes all the response fields?

Reviewer: I meant to reword this slightly (notice the word all):

```
The following table describes all the response fields:
```

| Statistic    | Description |
| :---         | :---        |
| `count`      | The number of documents sampled for the aggregation. |
| `mean`       | The average value of the field computed from the sample. |
| `variance`   | The square of deviation from the mean, a measure of data spread. |

> | [`skewness`](https://en.wikipedia.org/wiki/Skewness)   | A measure of the distribution's assymetry about the mean. |

Reviewer: Not sure if these stat names themselves should be clickable? I liked previous mode of saying, to learn more about xxxx see....

Author: Style guide specifies direct links in tables to save space.

          "skewness": 5.530161335032689,
          "kurtosis": 131.1630632404217,
          "covariance": {
            "products.base_price": 360.5035285833702,
            "taxful_total_price": 846.6489362233169
          },
          "correlation": {
            "products.base_price": 1,
            "taxful_total_price": 0.8444765264325269
          }
        },
        {
          "name": "taxful_total_price",
          "count": 4675,
          "mean": 75.05542864304839,
          "variance": 2788.1879749835425,
          "skewness": 15.812149139923994,
          "kurtosis": 619.1235507385886,
          "covariance": {
            "products.base_price": 846.6489362233169,
            "taxful_total_price": 2788.1879749835425
          },
          "correlation": {
            "products.base_price": 0.8444765264325269,
            "taxful_total_price": 1
          }
        }
      ]
    }
  }
}
```

The following table describes the response fields.

| Statistic    | Description |
| :---         | :---        |
| `count`      | The number of documents sampled for the aggregation. |
| `mean`       | The average value of the field computed from the sample. |

> | `variance`   | The square of deviation from the mean, a measure of data spread. |

Reviewer 3: Line 115: Is "about" the right word here?

Reviewer 2: It's understandable but I'll change to "relative to" to be more precise.

> | `skewness`   | A measure of the distribution's assymetry about the mean. See [Skewness](https://en.wikipedia.org/wiki/Skewness). |

Reviewer 3: ```suggestion
| `skewness`   | A measure of the distribution's asymmetry about the mean. See [Skewness](https://en.wikipedia.org/wiki/Skewness). |
```

> | `skewness`   | A measure of the distribution's assymetry about the mean. See [Skewness](https://en.wikipedia.org/wiki/Skewness). |

Reviewer 2: ```suggestion
| `skewness`   | A measure of the distribution's assymetry relative to the mean. See [Skewness](https://en.wikipedia.org/wiki/Skewness). |
```
